apis/python/src/tiledbsoma/io/ingest.py

```diff
@@ -2112,58 +2112,12 @@ def to_anndata(
     obsm = {}
     if "obsm" in measurement:
         for key in measurement.obsm.keys():
-            shape = measurement.obsm[key].shape
-            if len(shape) != 2:
-                raise ValueError(f"expected shape == 2; got {shape}")
-            if isinstance(measurement.obsm[key], DenseNDArray):
-                obj = cast(DenseNDArray, measurement.obsm[key])
-                matrix = obj.read().to_numpy()
-                # The spelling ``sp.csr_array`` is more idiomatic but doesn't exist until Python 3.8
-                obsm[key] = matrix
-            else:
-                # obsp is nobs x nobs.
-                # obsm is nobs x some number -- number of PCA components, etc.
-                matrix = measurement.obsm[key].read().tables().concat().to_pandas()
-                nobs_times_width, coo_column_count = matrix.shape
-                if coo_column_count != 3:
-                    raise SOMAError(
-                        f"internal error: expect COO width of 3; got {coo_column_count}"
-                    )
-                if nobs_times_width % nobs != 0:
-                    raise SOMAError(
-                        f"internal error: encountered non-rectangular obsm[{key}]: {nobs} does not divide {nobs_times_width}"
-                    )
-                obsm[key] = conversions.csr_from_tiledb_df(
-                    matrix, nobs, nobs_times_width // nobs
-                ).toarray()
+            obsm[key] = _extract_obsm_or_varm(measurement.obsm[key], "obsm", key, nobs)
```

Comment:
maybe pass measurement, "obsm", and key, and then do `soma_nd_array = measurement[collection_name][element_name]` in the `_extract_obsm_or_varm` function to reduce repetition?